### PR TITLE
feat: integration testing for postgres dialect

### PR DESCRIPTION
**Pull request by asthamohta**

### CODE REVIEW

File: system-test/spanner.ts [hunk 1]
```diff
@@ -252,37 +302,60 @@ describe('Spanner', () => {
...
+          params: {
+            a: data[0].Key,
+            b: data[1].Key,
+          },
+        };
+        if (dialect === Spanner.POSTGRESQL) {
+          table = postgreSqlTable;
+          database = PG_DATABASE;
+          query = {
+            sql: `SELECT * FROM ${table.name.toLowerCase()} WHERE "Key" = $1 OR "Key" = $2`,
```
Comment (olavloite) [inline system-test/spanner.ts:318]: nit: I don't think we need the `.toLowerCase()` here. The table name will be folded to lowercase automatically by the backend.

File: system-test/spanner.ts [hunk 2]
```diff
@@ -1050,23 +1369,45 @@ describe('Spanner', () => {
...
     const TABLE_NAME = 'SingersTest';
-    it('should auto create a database', done => {
-      const database = instance.database(generateName('database'));
+    const autoCreateDatabase = (done, databaseId) => {
+      const database = instance.database(generateName(databaseId));
 
       database.get({autoCreate: true} as GetDatabaseConfig, err => {
         assert.ifError(err);
         RESOURCES_TO_CLEAN.push(database);
         database.getMetadata(done);
```
Comment (olavloite) [inline system-test/spanner.ts:1378]: nit: Maybe add a check here to verify that the database was created with the correct dialect.

File: system-test/spanner.ts [hunk 3]
```diff
@@ -1116,37 +1457,63 @@ describe('Spanner', () => {
...
         execAfterOperationComplete(err => {
           assert.ifError(err);
 
-          function replaceNewLinesAndSpacing(str) {
-            return str.replace(/\n\s*/g, '').replace(/\s+/g, ' ');
+          function replaceNewLinesAndSpacing(str, dialect) {
+            const schema = str.replace(/\n\s*/g, '').replace(/\s+/g, ' ');
+            if (dialect === Spanner.GOOGLE_STANDARD_SQL) {
+              return schema;
+            } else return schema.toLowerCase();
```
Comment (olavloite) [inline system-test/spanner.ts:1470]: nit: add curly braces for the else statement

File: system-test/spanner.ts [hunk 4]
```diff
@@ -1221,77 +1610,155 @@ describe('Spanner', () => {
...
+          databaseDialect: Spanner.POSTGRESQL,
+          gaxOptions: GAX_OPTIONS,
+        });
+      await postgreSqlDatabase2CreateOperation.promise();
+
+      const schema = [
+        `CREATE TABLE Albums (
+            AlbumId VARCHAR NOT NULL PRIMARY KEY,
+        AlbumTitle VARCHAR NOT NULL
+    );`,
```
Comment (olavloite) [inline system-test/spanner.ts:1653]: nit: The formatting seems a little off here

File: system-test/spanner.ts [hunk 5]
```diff
@@ -1317,9 +1784,28 @@ describe('Spanner', () => {
...
+    it('GOOGLE_STANDARD_SQL should have completed a backup', async () => {
+      await completedBackup(
+        googleSqlBackup1,
+        googleSqlBackup1Name,
+        googleSqlDatabase1
+      );
+    });
+
+    it('POSTGRESQL should have completed a backup', async function () {
+      if (IS_EMULATOR_ENABLED) {
```
Comment (olavloite) [inline system-test/spanner.ts:1798]: nit: here and elsewhere, I think you can remove this, as backup tests should be disabled for both dialects on the emulator

**Review by olavloite — APPROVED** [review]
Changes LGTM (with a couple of nits)

This will add a lot of additional integration tests, which means that the build time will increase. Is there any indication on how much?

**Comment by asthamohta** [comment]
> Changes LGTM (with a couple of nits)
> 
> This will add a lot of additional integration tests, which means that the build time will increase. Is there any indication on how much?

Haven't looked into that but for golang we are skipping backup tests. What do you think?

File: system-test/spanner.ts (hunk 2 above)
Comment (asthamohta) [reply system-test/spanner.ts:1378]: Adding thanks

**Comment by olavloite** [comment]
> > Changes LGTM (with a couple of nits)
> > This will add a lot of additional integration tests, which means that the build time will increase. Is there any indication on how much?
> 
> Haven't looked into that but for golang we are skipping backup tests. What do you think?

I think that it would be good to skip backup tests for PostgreSQL. I don't think they add enough value to justify the additional build time (the last successful build with all the tests in this PR was 52 minutes, I think)

**Pull request merged**